Doc/library/typing.rst

```diff
@@ -2804,6 +2804,47 @@ Introspection helpers
 
    .. versionadded:: 3.8
 
+.. function:: get_orig_bases(tp, /)
+
+    Returns the objects in the bases list in the class's definition before
+    they were modified by ``__mro_entries__``. This is useful for
+    introspecting ``Generic``\s.
+
+    Examples::
+
+        T = TypeVar("T")
+
+        class Foo(Generic[T]): ...
+
+        class Bar(Foo[int], float): ...
+
+        get_orig_bases(Foo) == (Generic[T],)
+        get_orig_bases(Bar) == (Foo[int], float)
+        get_orig_bases(int) == None
+
+    .. versionadded:: 3.12
+
+.. function:: get_orig_class(tp, /)
+
+    Returns the ``GenericAlias`` object that was instantiated to create ``tp``.
+
+    Examples::
+
+        T = TypeVar("T")
+
+        class Foo(Generic[T]): ...
+
+        get_orig_class(Foo[int]()) == Foo[int]
+        get_orig_class(list[int]()) == None
```

Comment:
If this function won't work for `types.GenericAlias`, I'm not sure it's useful enough to add to the standard library.

Reply:
It only doesn't work if the class is "slotted" and doesn't include `__orig_class__` but if the class has a `__dict__` or includes the slot it will work. (this is the same requirement for the `typing._GenericAlias` version to work as well)
```py
In [1]: from types import GenericAlias

In [2]: class Foo:
   ...:     __class_getitem__ = classmethod(GenericAlias)
   ...: 

In [3]: Foo[int]().__orig_class__
Out[3]: __main__.Foo[int]
```

Reply:
I see. Then I feel we should not add this unless we make it work for `list[int]`. That might require changes deep in the core, so not sure we'll be able to do that.

Reply:
I agree with @JelleZijlstra that if this function is inconsistent between `types.GenericAlias` and `typing._GenericAlias`, it seems like it could be too confusing to be worth adding.

Reply:
I'm assuming I'd need a PEP to add a new slot member to the PyObject struct

Reply:
> I'm assuming I'd need a PEP to add a new slot member to the PyObject struct

I'd guess so, yeah

Reply:
I guess we document the attribute at the least somewhere until then

Reply:
> I guess we document the attribute at the least somewhere until then

Maybe. Could we possibly tackle that in a separate PR, rather than documenting it in the same PR as adding `types.get_orig_bases`?

Reply:
I don't think we'd need a PEP, it's just an internal implementation detail where we put these things. Worth going through the implications carefully, though.

Reply:
I'll make a post on the python discourse but I feel like that might take a while to go through and it's probably better off as a separate PR.